rust/lance-linalg/src/kmeans.rs

```diff
@@ -545,25 +540,41 @@ fn get_slice(data: &[f32], x: usize, y: usize, dim: usize, strip: usize) -> &[f3
     &data[x * dim + y..x * dim + y + strip]
 }
 
+fn compute_partitions_l2_f32_small<'a>(
+    centroids: &'a [f32],
+    data: &'a [f32],
+    dim: usize,
+) -> Box<dyn Iterator<Item = (u32, f32)> + 'a> {
+    Box::new(data.chunks(dim).map(move |row| {
+        argmin_value(centroids.chunks(dim).map(|centroid| row.l2(centroid))).unwrap()
+    }))
+}
+
 /// Fast partition computation for L2 distance.
-fn compute_partitions_l2_f32(centroids: &[f32], data: &[f32], dim: usize) -> Vec<u32> {
+fn compute_partitions_l2_f32<'a>(
+    centroids: &'a [f32],
+    data: &'a [f32],
+    dim: usize,
+) -> Box<dyn Iterator<Item = (u32, f32)> + 'a> {
```

Comment:
Since we are throwing data away later with `.map(|(c, _)| c)` (line 654), it would be nice to not Box this, since boxing prevent optimization. I think you could achieve by making these return non-boxed iterators:

```suggestion
) -> impl Iterator<Item = (u32, f32)> + 'a {
```

And then use [Either](https://docs.rs/either/latest/either/enum.Either.html) to wrap the two returned iterators. So,

```rust
    if std::mem::size_of_val(centroids) <= 16 * 1024 {
        return Either::Left(compute_partitions_l2_f32_small(centroids, data, dim));
    }
```

and

```rust
Either::right(stream)
```

Reply:
sg. lemme try it.

Reply:
That not work?

Reply:
What if we wanted to return 3 different streams ?

Reply:
> What if we wanted to return 3 different streams ?

`Either` is just an enum.  If there were three possible returns then you would need your own enum in that case.

Reply:
Yea i mean this is not significantly better than a `Box<&dyn Iterator>` then?

Reply:
> Yea i meant this is not significantly better than a Box<&dyn Iterator> then?

`Box<&dyn Iterator>` creates a heap allocation and prevents optimization (since it brings it back to dynamic dispatch).
`Either` (or another enum) is still stack allocation and can be optimized by the compiler.

Reply:
Here's a simple benchmark showing the dyn / impl tradeoff: https://gist.github.com/westonpace/e919fc5716c78dde929ec572b87bb465

```
test result: ok. 0 passed; 0 failed; 0 ignored; 0 measured; 0 filtered out; finished in 0.00s

     Running benches/main.rs (target/release/deps/main-b648dcdeee0e3988)
impl                    time:   [137.54 ns 137.85 ns 138.21 ns]
dyn                     time:   [4.3391 µs 4.3482 µs 4.3579 µs]
```

Reply:
Here's a version using either, which behaves pretty much the same as impl directly.  So the enum penalty is much lower:

https://gist.github.com/westonpace/410434e371c27b276de636f6486dff0a